numba/np/arrayobj.py

```diff
@@ -4958,12 +4958,93 @@ def np_flip(a):
         raise errors.TypingError("Cannot np.flip on %s type" % a)
 
     def impl(a):
-        sl = _build_slice_tuple(a.ndim)
+        sl = _build_flip_slice_tuple(a.ndim)
         return a[sl]
 
     return impl
 
 
+@intrinsic
+def _build_full_slice_tuple(tyctx, sz):
+    """Creates a sz-tuple of full slices"""
+    size = int(sz.literal_value)
+    tuple_type = types.UniTuple(dtype=types.slice2_type, count=size)
+    sig = tuple_type(sz)
+
+    def codegen(context, builder, signature, args):
+        def impl(length, empty_tuple):
+            out = empty_tuple
+            for i in range(length):
+                out = tuple_setitem(out, i, slice(None, None))
+            return out
+
+        inner_argtypes = [types.intp, tuple_type]
+        inner_sig = typing.signature(tuple_type, *inner_argtypes)
+        ll_idx_type = context.get_value_type(types.intp)
+        # Allocate an empty tuple
+        empty_tuple = context.get_constant_undef(tuple_type)
+        inner_args = [ll_idx_type(size), empty_tuple]
+
+        res = context.compile_internal(builder, impl, inner_sig, inner_args)
+        return res
+
+    return sig, codegen
+
+
+@overload(np.array_split)
+def np_array_split(ary, indices_or_sections, axis=0):
+    # If this statement is put at the top of the file, numba fails to import
+    from numba.typed import List
+
+    if ary.ndim > 1:
+        a_type = types.Array(ary.dtype, ary.ndim, "A")
+    else:
+        a_type = typeof(ary)
+
+    if isinstance(indices_or_sections, types.Integer):
+
+        def impl(ary, indices_or_sections, axis=0):
+            l, rem = divmod(ary.shape[axis], indices_or_sections)
+            return np.array_split(
+                ary, np.arange(l + rem, ary.shape[axis], l), axis=axis
+            )
+
+    else:
```

Comment:
I think this ought to strongly restrict what types other than `Integer` are permissible to prevent strange errors from unhandled types being used.

Reply:
To me, this should just be an iterable (preferably of known size, though this isn't a strict requirement) of integers. How could I type that?

Reply:
Would `isinstance(<arg>, numba.types.IterableType)` work?

Reply:
I'm not completely sure. It looks like it might, but I'm not sure how these instance checks work. Would anything that passed that check have a `yield_type` attribute?

I'm thinking this could be the check:

```python
    elif (
        isinstance(indices_or_sections, types.IterableType) 
        and isinstance(indices_or_sections.iterator_type.yield_type, types.Integer)
    ):
```

Reply:
Hmm. `UniTuple` works, but `Tuple` does not. Not sure if this would be an issue, but it does disallow cases like: `np.split(a, (np.int32(1), np.int64(4)))`

Reply:
I suspect that this is because iteration requires the induction variable to be the same type across iterations, else the loop body has to be versioned, see [`literal_unroll`](http://numba.pydata.org/numba-doc/latest/reference/pysupported.html?highlight=literal_unroll#heterogeneous-tuples).